runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+        // way each batch will have unique queries. If we coalesce after batching, then its possible to have common
+        // stream queries across batches, this means that the same stream is queried multiple times, but it could
+        // have only been polled once.
+
+        List<StreamAddressRange> coalesced = coalesce(queries);
+        List<List<StreamAddressRange>> batches = Lists.partition(coalesced, pollBatchSize);
+
+        Map<UUID, StreamAddressSpace> allQueryResults = new HashMap<>();
+
+        for (List<StreamAddressRange> batch : batches) {
+            long pollStartTime = System.nanoTime();
+            Map<UUID, StreamAddressSpace> res = sequencerView.getStreamsAddressSpace(coalesced);
+            MicroMeterUtils.time(Duration.ofNanos(System.nanoTime() - pollStartTime), "StreamPollingScheduler.poll");
+            Preconditions.checkState(res.size() == batch.size());
+            allQueryResults.putAll(res);
+        }
+
+        Preconditions.checkState(tasks.size() == queries.size());
+
+        for (int idx = 0; idx < tasks.size(); idx++) {
+            StreamingTask task = tasks.get(idx);
+            StreamAddressRange taskQuery = queries.get(idx);
+            Preconditions.checkState(task.getStream().getStreamId().equals(taskQuery.getStreamID()));
+            Preconditions.checkState(allQueryResults.containsKey(taskQuery.getStreamID()),
+                    "StreamAddressSpace missing for %s", task.getStream().getStreamId());
+
+            StreamAddressSpace sas = allQueryResults.get(task.getStream().getStreamId()).getAddressesInRange(taskQuery);
+            task.getStream().refresh(sas);
+        }
+    }
+
+    @Data
+    class Interval {
+        private final long x;
+        private final long y;
+    }
+
+    private Interval merge(Interval interval, StreamAddressRange range) {
+        long newMinX = Math.min(interval.x, range.getEnd());
+        long newMaxY = Math.max(interval.y, range.getStart());
+        return new Interval(newMinX, newMaxY);
+    }
+
+    // Generate a range per stream that contains all sub-ranges that need to be queries
+
+    /**
+     * It's possible that multiple streaming tasks need to query different/overlapping ranges
+     * of the same stream. In order to prevent issuing multiple queries for the same stream,
+     * all range queries for the same stream are coalesced into a single request.
+     * For example, the following queries will all be merged into one query (i.e., (-1, 100])
+     *  Query1  (-1, 100)
+     *  Query2  (2, 5)
+     *  Query3  (3, 10)
+     */
+    private List<StreamAddressRange> coalesce(List<StreamAddressRange> ranges) {
+        Map<UUID, Interval> adjusted = new HashMap<>();
+
+        for (StreamAddressRange range : ranges) {
+            Interval interval = adjusted.getOrDefault(range.getStreamID(), new Interval(Long.MAX_VALUE, Long.MIN_VALUE));
```

Comment:
To me that seems like a specific use case, but shouldn't be enforced by the StreamingManager itself because it can restrict other use cases. Some of our tests actually register two different listener on the same tag. Anyways, the overhead of this is coalescing logic is negligible (based on profiling) for the case you mentioned. 

I think its better to track the number of sequence calls the scheduler emits per scheduling tick, that metric will reflect how efficiently we are batching.